### PR TITLE
fix: correct NameError by moving import out of type checking clause 

### LINKED ISSUE #384: `commands.Param` can't be used with builtin converter types
### Reproduction Steps

register a command like below, and try to invoke it

### Minimal Reproducible Code

```python
@bot.slash_command()
async def test(
    inter: disnake.CommandInteraction,
    message: disnake.Message = commands.Param(),
):
    await inter.send(repr(message))
```

### Expected Results

it converts the parameter (e.g. a message url) to a `disnake.Message` and passes it to the command handler

### Actual Results

```py
Ignoring exception in slash command 'test':
Traceback (most recent call last):
  File "...\disnake\ext\commands\slash_core.py", line 597, in invoke
    await call_param_func(self.callback, inter, self.cog, **kwargs)
  File "...\disnake\ext\commands\params.py", line 767, in call_param_func
    cog_param, inter_param, paraminfos, injections = collect_params(function)
  File "...\disnake\ext\commands\params.py", line 701, in collect_params
    paraminfo = ParamInfo.from_param(parameter, {}, doc)
  File "...\disnake\ext\commands\params.py", line 385, in from_param
    self.parse_annotation(type_hints.get(param.name, param.annotation))
  File "...\disnake\ext\commands\params.py", line 478, in parse_annotation
    self.parse_converter_annotation(self.converter, annotation)
  File "...\disnake\ext\commands\params.py", line 536, in parse_converter_annotation
    _, parameters = isolate_self(converter)
  File "...\disnake\ext\commands\params.py", line 642, in isolate_self
    sig = signature(function)
  File "...\disnake\ext\commands\params.py", line 126, in signature
    typehints = get_type_hints(func)
  File "...\Python39\lib\typing.py", line 1496, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "...\Python39\lib\typing.py", line 292, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "...\Python39\lib\typing.py", line 554, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'AnyContext' is not defined

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "...\disnake\ext\commands\interaction_bot_base.py", line 1332, in process_application_commands
    await app_command.invoke(interaction)
  File "...\disnake\ext\commands\slash_core.py", line 606, in invoke
    raise CommandInvokeError(exc) from exc
disnake.ext.commands.errors.CommandInvokeError: Command raised an exception: NameError: name 'AnyContext' is not defined
```

### Intents

irrelevant

### System Information

master, py 3.9

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

this has been happening for a while (with `Context` instead of `AnyContext`) and was just never really fixed

### PR DESCRIPTION
**Pull request by bast0006**
## Summary

This PR fixes #384, correcting an import cycle as well as adding support for Union[] inside the self and cog extractor function `isolate_self`. 

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue (#384)
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/ext/commands/context.py [hunk 1]
```diff
@@ -32,6 +32,8 @@
 import disnake.utils
 from disnake.message import Message
 
+from ... import ApplicationCommandInteraction
```
Comment (onerandomusername) [inline disnake/ext/commands/context.py:35]: probably should use `from disnake import ApplicationCommandInteraction`

**Review by shiftinv — COMMENTED** [review]
Looks good, thanks!
I don't think this solves the underlying issue of the converter annotations only being parsed when the command is invoked, and only if `commands.Param` is used. The reason for that appears to be the fact that the converter parsing happens before the converter gets set, and since the parameter parsing results are only persistent between calls when using `commands.Param`, the annotation only gets parsed when the command is invoked.
Anyhow, I think this'll do, since the builtin converters all accept the default type (string) so the annotations of those don't really have to be parsed.

**Review by onerandomusername — APPROVED** [review]
lgtm

File: disnake/ext/commands/params.py [hunk 2]
```diff
@@ -95,7 +96,15 @@
...
+    if not isinstance(obj, type):
+        # Assume we have a type hint
+        if get_origin(obj) in (Union, UnionType, Optional):
+            obj = get_args(obj)
+            return any(issubclass(o, tp) for o in obj)
+        else:
+            # Other type hint specializations are not supported
+            return False
+    elif not isinstance(tp, (type, tuple)):
         return False
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/ext/commands/params.py]: https://github.com/DisnakeDev/disnake/pull/398/commits/410a7d049c893ccc7ed30a00409661e6fd7c5d13

File: disnake/ext/commands/converter.py [hunk 3]
```diff
@@ -48,14 +48,13 @@
...
 import disnake
 
-from .context import Context
+from .context import AnyContext, Context
 from .errors import *
 
 if TYPE_CHECKING:
     from disnake.message import MessageableChannel
 
     # TODO: don't use unbound generic `Context`
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/ext/commands/converter.py]: https://github.com/DisnakeDev/disnake/pull/398/commits/410a7d049c893ccc7ed30a00409661e6fd7c5d13

File: disnake/ext/commands/context.py (hunk 1 above)
Comment (onerandomusername) [reply disnake/ext/commands/context.py]: https://github.com/DisnakeDev/disnake/pull/398/commits/410a7d049c893ccc7ed30a00409661e6fd7c5d13

**Pull request merged**